Add part of the jit-vmap doc. (#188)
EMI-Group/evox: docs/source/guide/developer/2-jit-vmap-toturial-xbq.md

## How to use JIT for a subclass of ModuleBase

For better understanding of this part, we need to explain 3 important functions in EvoX: [`jit_class`](#evox.core.module.jit_class), [`vmap`](#evox.core.jit_util.vmap) and [`jit`](l#evox.core.jit_util.jit).

### jit_class function

[`jit_class`](#evox.core.module.jit_class) is a helper function used to Just-In-Time (JIT) script of Pytorch ([`torch.jit.script`](https://pytorch.org/docs/stable/generated/torch.jit.script.html)) or trace ([`torch.jit.trace_module`](https://pytorch.org/docs/stable/generated/torch.jit.trace_module.html#torch-jit-trace-module)) all member methods of the input class. 

[`jit_class`](#evox.core.module.jit_class) has two parameters:

- `cls`: the original class whose member methods are to be lazy JIT.
- `trace`: whether to trace the module or to script the module. Default to `False`.

```{note}

> 1.In many cases, it is not necessary to wrap your custom algorithms or problems with [`jit_class`](#evox.core.module.jit_class), the workflow(s) will do the trick for you.

Reviewer: you forgot the space after `1.`
```suggestion
1. In many cases, it is not necessary to wrap your custom algorithms or problems with [`jit_class`](#evox.core.module.jit_class), the workflow(s) will do the trick for you.
```

2.With `trace=True`, all the member functions are effectively modified to return `self` additionally since side-effects cannot be traced. If you want to preserve the side effects, please set `trace=False` and use the `use_state` function to wrap the member method to generate pure-functional (the `use_state` function will be explained in the next part).
3.Similarly, all module-wide operations like `self.to(...)` can only returns the unwrapped module, which may not be desired. Since most of them are in-place operations, a simple `module.to(...)` can be used instead of `module = module.to(...)`.
```

### vmap function

[`vmap`](#evox.core.jit_util.vmap) function vectorized map the given function to its mapped version. Based on [`torch.vmap`](https://pytorch.org/docs/main/generated/torch.vmap.html), we made many improvements, and you can see [`torch.vmap`](https://pytorch.org/docs/main/generated/torch.vmap.html) for more information.

### jit function

> [`jit`](l#evox.core.jit_util.jit) compile the given `func` via [`torch.jit.trace`](https://pytorch.org/docs/stable/generated/torch.jit.script.html) (`trace=True`) and [`torch.jit.script`](https://pytorch.org/docs/stable/generated/torch.jit.trace.html) (`trace=False`).

Reviewer: typo here
```suggestion
[`jit`](#evox.core.jit_util.jit) compile the given `func` via [`torch.jit.trace`](https://pytorch.org/docs/stable/generated/torch.jit.script.html) (`trace=True`) and [`torch.jit.script`](https://pytorch.org/docs/stable/generated/torch.jit.trace.html) (`trace=False`).
```

  This function wrapper effectively deals with nested JIT and vector map (`vmap`) expressions like `jit(func1)` -> `vmap` -> `jit(func2)`, preventing possible errors.

```{note}
1. With `trace=True`, `torch.jit.trace` cannot use SAME example input arguments for function of DIFFERENT parameters,e.g., you cannot pass `tensor_a, tensor_a` to `torch.jit.trace`d version of `f(x: torch.Tensor, y: torch.Tensor)`.
2. With `trace=False`, `torch.jit.script` cannot contain `vmap` expressions directly, please wrap them with `jit(..., trace=True)` or `torch.jit.trace`.
```

> In the [Working with Module in EvoX](#/guide/developer/1-modulebase), we have briefly introduced some rules about the methods inside a subclass of the [`ModuleBase`](#evox.core.module.ModuleBase) . Now that [`jit_class`](#evox.core.module.jit_class), [`vmap`](#evox.core.jit_util.vmap) and [`jit`](l#evox.core.jit_util.jit) have been explianed,  we will explain more rules and provide some specific hints.

Reviewer: same here

    
    ...
    
```

```{note}
If method A invokes vmapped method B, then A and all methods invoke method A can not be vmapped again.
```

### Internal vmapped methods inside the subclass

Inside the subclass, internal vmapped methods can be  JIT by using the [`trace_impl`](#evox.core.module.trace_impl):

```Python
# Import Pytorch
import torch

# Import the ModuleBase class from EvoX
from evox.core import ModuleBase, jit, vmap, trace_impl

# Set an module inherited from the ModuleBase class
class ExampleModule(ModuleBase):
    
    ...    
    
    # The internal vmapped class method to be JIT   
    @jit
    def jit_vamp_func(self, p: torch.Tensor) -> torch.Tensor:
        
        # The original method
        # We can not vmap it
    	def func(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
    		return x+y
        
        
        # The method to be vmapped
        # We need to use trace_impl to rewrite the original method
        @trace_impl(func)
    	def trace_func(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
    		pass
        
        return vmap(func, in_dims=1, out_dims=1, trace=False)(p, p)
    
    ...
    
```

```{note}

> If a class method use [`trace_impl`](#evox.core.module.trace_impl),it will be only available in the trace mode. More details about `[`trace_impl`](#evox.core.module.trace_impl) will be shown in the next part.

Reviewer: You don't need that many references in the same place. :)
```suggestion
If a class method uses [`trace_impl`](#evox.core.module.trace_impl), it will be only available in the trace mode. More details about trace_impl will be provided in the next section.
```
